src/coalition_evaluator.py

```diff
@@ -0,0 +1,224 @@
+    num_voters = voting_data.shape[1]
+    successful_coalitions = []
+    evaluated_scenarios = set()  # Keep track of evaluated scenarios to avoid duplicates
+
+    # Original setup
+    original_winner = plurality_outcome(voting_data).winner
+    print(voting_data)
+    print("original winner: " + original_winner + "\n")
+    original_happiness = HappinessLevel(voting_data, original_winner)
+
+    # Analyzing possible coalitions
+    for r in range(2, num_voters + 1):
+        for initial_coalition in combinations(range(num_voters), r):
+            for candidate in set(voting_data[0, :]):  # Potential strategic votes
+                # Unique identifier for a coalition scenario based on its members and the strategic candidate
+                coalition_key = (tuple(sorted(initial_coalition)), candidate)
+                # Check if this scenario was already evaluated
+                if coalition_key in evaluated_scenarios:
+                    continue  # Skip if already evaluated
+
+                new_voting_array = voting_data.copy()  # Make a new array for changes
+                actual_coalition = []  # Track actual members who changed their vote
+
+                for member in initial_coalition:
+                    if candidate in new_voting_array[:, member]:
+                        idx = np.where(new_voting_array[:, member] == candidate)[0][0]
+                        if idx != 0:  # Only consider if candidate is not already the top choice
+                            new_voting_array[:, member] = np.roll(new_voting_array[:, member], -idx)
+                            actual_coalition.append(member)  # Add to actual coalition
+
+                if actual_coalition:
+                    new_winner = plurality_outcome(new_voting_array).winner
+                    new_happiness = HappinessLevel(voting_data, new_winner)
+
+                    # Check if all actual coalition members are happier with the new outcome
+                    if all(new_happiness.happiness_level_dict[f"voter_{member}"] > original_happiness.happiness_level_dict[f"voter_{member}"] for member in actual_coalition):
+                        successful_coalitions.append((actual_coalition, candidate, new_winner))
+                        evaluated_scenarios.add(coalition_key)  # Mark this scenario as evaluated
+                        print(new_voting_array)
+                        print(f"Successful Coalition\n: {actual_coalition}, Strategic Vote: {candidate}, New Winner: {new_winner}")
+                        print("Changes in Happiness Levels for Coalition Members:")
+                        for member in actual_coalition:
+                            happiness_diff = new_happiness.happiness_level_dict[f"voter_{member}"] - original_happiness.happiness_level_dict[f"voter_{member}"]
+                            print(f"Voter {member}: Happiness Change = {happiness_diff:.3f}")
+                        print("\n")
+
+    return successful_coalitions
```

Comment:
In the assignment description, it asks to compute 𝑠𝑖𝑗 = (𝑣̃𝑖𝑗, 𝑂̃ , 𝐻̃𝑖 , 𝐻𝑖, 𝐻̃ , 𝐻) for each strategic voting scenario. Additionally, take a look at the output provided by 'strategic_voting.py', which is a Pandas DataFrame containing the previous values (as in the notebook).